contributors/design-proposals/auth/bound-service-account-tokens.md

```diff
@@ -0,0 +1,180 @@
+# Bound Service Account Tokens
+
+Author: @mikedanese
+
+# Objective
+
+This document describes an API that would allow workloads running on Kubernetes
+to request JSON Web Tokens that are audience, time and eventually key bound.
+
+# Background
+
+Kubernetes already provisions JWTs to workloads. This functionality is on by
+default and thus widely deployed. The current workload JWT system has serious
+issues:
+
+1.  Security: JWTs are not audience bound. Any recipient of a JWT can masquerade
+    as the presenter to anyone else.
+1.  Security: The current model of storing the service account token in a Secret
+    and delivering it to nodes results in a broad attack surface for the
+    Kubernetes control plane when powerful components are run - giving a service
+    account a permission means that any component that can see that service
+    account's secrets is at least as powerful as the component.
+1.  Security: JWTs are not time bound. A JWT compromised via 1 or 2, is valid
+    for as long as the service account exists. This may be mitigated with
```

Comment:
non-expiring tokens can be revoked by removing the associated Secret or ServiceAccount object.

I think we'd still want some form of this in order to be able to generate legacy-compatible tokens out of the API. What about this:
* a non-expiring token must be bound to a deletable object name+uid (like a pod or secret)
* non-expiring tokens can only be for the apiserver audience (since only the apiserver can be counted on to verify the bound object still exists)
* the API server must verify the object+uid still exists before verifying the token

This gets the confidential and bulky portion of unexpiring service account tokens out of the API, and still lets them be revocable. It also lets us move to this new model for both legacy and expiring/rotating cases, which is desirable.

Reply:
Not only desirable, required.  I think we want to completely supplant the old model.

Reply:
This makes key OIDC signing key rotation difficult or impossible. Do we need to support non-expiring forever or just for some number of releases (in accordance to our GA deprecation policy).

Reply:
> This makes key OIDC signing key rotation difficult or impossible.

How so? A non-expiring token doesn't mean it can't be revoked or made invalid, it just doesn't have an inherent expiration time.

Reply:
It means you could never remove a public key form the JWKS_URI since it could potentially need to exist forever to verify a token.

Reply:
I'd consider that working as intended. No inherent expiration, but an admin decision to revoke or rotate signing keys could invalidate it.

Reply:
The point is that it's not easy to maintain compatibility with the current client-go while also rotating SA signing keys. If we rotate signing keys every 4 days, how is a non-expiring token useful?